libs/core/src/main/java/com/sourcefuse/jarc/core/entitylisteners/AuditLogEntityListener.java

```diff
@@ -0,0 +1,88 @@
+package com.sourcefuse.jarc.core.entitylisteners;
+
+import com.google.gson.Gson;
+import com.sourcefuse.jarc.core.awares.ApplicationAwareBeanUtils;
+import com.sourcefuse.jarc.core.constants.Constants.AuditActions;
+import com.sourcefuse.jarc.core.models.audit.AuditLog;
+import com.sourcefuse.jarc.core.models.base.BaseEntity;
+import com.sourcefuse.jarc.core.models.base.SoftDeleteEntity;
+import com.sourcefuse.jarc.core.models.session.CurrentUser;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.PostPersist;
+import jakarta.persistence.PostRemove;
+import jakarta.persistence.PostUpdate;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+@Slf4j
+public class AuditLogEntityListener<T extends BaseEntity> {
+
+  @PostPersist
+  public void onPersist(T target) {
+    this.saveAuditLog(AuditActions.SAVE, target);
+  }
+
+  @PostUpdate
+  public void onUpdate(T target) {
+    if (target instanceof SoftDeleteEntity) {
+      SoftDeleteEntity softDeletedEntity = (SoftDeleteEntity) target;
+      if (!softDeletedEntity.isDeleted()) {
+        this.saveAuditLog(AuditActions.UPDATE, target);
+      } else {
+        this.saveAuditLog(AuditActions.DELETE, target);
+      }
+    } else {
+      this.saveAuditLog(AuditActions.UPDATE, target);
+    }
+  }
+
+  @PostRemove
+  public void onRemove(T target) {
+    this.saveAuditLog(AuditActions.DELETE, target);
+  }
+
+  private EntityManager getEntityManager() throws Exception {
+    EntityManager em = ApplicationAwareBeanUtils.getBean(EntityManager.class);
+    if (em == null) {
+      throw new Exception(
+        "Entity Manager is null can not proceed to save audit log"
+      );
+    }
+    return em.getEntityManagerFactory().createEntityManager();
+  }
+
+  @SuppressWarnings("unchecked")
+  private void saveAuditLog(AuditActions action, T entity) {
+    try {
+      String before = null;
+      String after = new Gson().toJson(entity);
+      EntityManager em = this.getEntityManager();
+      if (!action.toString().contains("SAVE")) {
+        T oldEntity = (T) em.find(entity.getClass(), entity.getId());
+        before = oldEntity != null ? new Gson().toJson(oldEntity) : null;
+      }
+      if (action.toString().contains("DELETE")) {
+        after = null;
+      }
+      em.getTransaction().begin();
+      CurrentUser<?> currentUser = (CurrentUser<?>) SecurityContextHolder
+        .getContext()
+        .getAuthentication()
+        .getPrincipal();
+      AuditLog auditLog = new AuditLog();
+      auditLog.setAction(action);
+      auditLog.setActedOn(entity.getTableName());
+      auditLog.setActionKey(entity.getClass().getSimpleName() + "_Logs");
+      auditLog.setBefore(before);
+      auditLog.setAfter(after);
+      auditLog.setEntityId(entity.getId());
+      auditLog.setActor(currentUser.getUser().getId());
+      log.info("::: audit Log {}", auditLog.toString());
```

Comment:
which logger is this ? Can we standardize a log format for all logs ?

Reply:
logger is slf4j, can you provide the log format will change the format for all logs